### PR TITLE
Mention if missing symbol is on the class path

### LINKED ISSUE #12289: Clashing type names across SBT modules doesn't result in an error
## reproduction steps
scala version: `2.12.10`

I have two SBT modules `A` and `B`
Module A defines some classes

```
package a.b.c
abstract class ResourceType[T](val amount: T) {
}

class Cpu(amount: Double) extends ResourceType[Double](amount){
}

class Memory(amount: ByteSize) extends ResourceType[ByteSize](amount){
}


class Storage(amount: ByteSize) extends ResourceType[ByteSize](amount){
}

class RuntimeResources(cpus: Double = 1, memSize: ByteSize = 0.B, storageSize: ByteSize = 0.B) {
  val cpu: Cpu = new Cpu(cpus)
  val mem: Memory = new Memory(memSize)
  val storage: Storage = new Storage(storageSize)
}
```

Then in module `B` I have a file within the same package a trait of the same name as above (but a different SBT module)

```
package a.b.c
sealed trait TestTrait
case object Memory extends TestTrait
```

Then in another file within module B I have
```
...
val x: Seq[RuntimeResources] = ...
...
x.map(_.mem.amount.toMB)
```

When I try to compile all of the modules I get the compilation error 

```
value amount is not a member of a.b.c.Memory
```

## problem

Note we have a clash in names within the same package `a.b.c` Two separate files across separate SBT modules define the `Memory` type. One is a trait, the other a class. At compile time, I would expect an error to be thrown about the name clash, rather than telling me that `amount` is not a member of `Memory`. This would indicate to me there is no issue with the naming of the class (or trait).

If I am unaware of the `Memory` trait the current error provides no useful information and is confusing to the user. This is particularly a problem for larger code bases where a developer might not know the existence of every single type.

### PR DESCRIPTION
**Pull request by som-snytt**
Hint if there is an object in scope, in a package, which has a class, and the class path finds it.

Fixes scala/bug#12289

### CODE REVIEW

**Comment by som-snytt** [comment]
`java.io.IOException: No space left on device`

**Comment by SethTisue** [comment]
I re-ran the Jenkins build, now that the community build runs that were using all the disk have finished.

I wonder sometimes if we should have a dedicated worker behemoth just for PR validation, separate from the behemoths used for the community build.

**Comment by som-snytt** [comment]
Needs refinement, as it does add verbiage for incremental compilation error.

```
[error] /home/amarki/snips/test-12289/src/main/scala/c.scala:4:17: not found: type T
[error] A definition on the class path may be hidden by a source artifact. (Companions must be defined together.)
```
because (as will surprise no one) the class is still there, i.e., previous artifacts are not cleaned first.
```
./target/scala-2.13/classes/p/Main.class
./target/scala-2.13/classes/p/T.class
./target/scala-2.13/classes/p/Main$.class
./target/scala-2.13/classes/p/C.class
./target/scala-2.13/classes/p/Main$delayedInit$body.class
```

**Comment by som-snytt** [comment]
The clever test is tweaked to allow `-d`.

The output check in typer may require tweaking: it uses single output only, and compares the path strings to mean "includes".

File: test/files/neg/t12289/c_2.scala [hunk 1]
```diff
@@ -0,0 +1,6 @@
+//> using options -d /tmp
```
Comment (lrytz) [inline test/files/neg/t12289/c_2.scala]: Maybe make it a `DirectTest` and use `testOutput.path`?

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala [hunk 2]
```diff
@@ -5669,7 +5669,25 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
           case LookupInaccessible(symbol, msg) => issue(AccessError(tree, symbol, context, msg))
           case LookupNotFound =>
             asTypeName orElse inEmptyPackage orElse lookupInRoot(name) match {
-              case NoSymbol => issue(SymbolNotFoundError(tree, name, context.owner, startContext, mode.in(all = PATTERNmode, none = APPSELmode | TYPEPATmode)))
+              case NoSymbol =>
+                def hasOutput = settings.outputDirs.getSingleOutput.isDefined || settings.outputDirs.outputs.nonEmpty
+                val hidden = !unit.isJava && name.isTypeName && hasOutput && {
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala]: The issue doesn't manifest the other way around, if `name.isTermName`?

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala [hunk 3]
```diff
@@ -5669,7 +5669,25 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
...
+                  startContext.lookupSymbol(name.toTermName, _ => true) match {
+                    case LookupSucceeded(qualifier, symbol) if symbol.owner.hasPackageFlag && symbol.sourceFile != null =>
+                      symbol.owner.info.decls.lookupAll(name).toList match {
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala:5682]: This I don't immediately understand; lookup for `name` failed (we're about to issue an error), but `lookupSymbol(name.toTermName).owner.info.decls.lookupAll(name)` gives a result. How is that?

File: src/compiler/scala/tools/nsc/typechecker/ContextErrors.scala [hunk 4]
```diff
@@ -330,9 +330,10 @@ trait ContextErrors extends splain.SplainErrors {
       def AmbiguousIdentError(tree: Tree, name: Name, msg: String) =
         NormalTypeError(tree, "reference to " + name + " is ambiguous;\n" + msg)
 
-      def SymbolNotFoundError(tree: Tree, name: Name, owner: Symbol, startingIdentCx: Context, inPattern: Boolean) = {
-        def help = if (inPattern && name.isTermName) s"\nIdentifiers ${if (name.charAt(0).isUpper) "that begin with uppercase" else "enclosed in backticks"} are not pattern variables but match the value in scope." else ""
-        NormalTypeError(tree, s"not found: ${decodeWithKind(name, owner)}$help")
+      def SymbolNotFoundError(tree: Tree, name: Name, owner: Symbol, inPattern: Boolean, hidden: Boolean) = {
+        val help = if (inPattern && name.isTermName) s"\nIdentifiers ${if (name.charAt(0).isUpper) "that begin with uppercase" else "enclosed in backticks"} are not pattern variables but match the value in scope." else ""
+        val path = if (hidden) s"\nA definition on the class path may be hidden by a source artifact. (Companions must be defined together.)" else ""
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/typechecker/ContextErrors.scala]: Maybe we can find a message that's a bit more explicit, hands-on. Connecting the dots between a "definition hidden by a source artifact" and "companions" is not easy simple.

**Review by lrytz — COMMENTED** [review]
It seems a helpful addition. Given

```
package p {
  object T
  class C extends T
}
```

and
```
package p {
  trait T
}
```

I assume IntelliJ would happily navigate from `extends T` to the definition of `trait T`, which makes it harder to figure why compilation fails.

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala (hunk 2 above)
Comment (som-snytt) [reply src/compiler/scala/tools/nsc/typechecker/Typers.scala]: I was too lazy to check after getting a first cut. Also, I considered it more likely to add an object as factory for an existing class.

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala (hunk 3 above)
Comment (som-snytt) [reply src/compiler/scala/tools/nsc/typechecker/Typers.scala:5682]: The package is loaded from class path first, which creates symbols of existing class + its module & module class. Then compilation of source takes over the existing module symbol, which is entered into scope.

It's fishy that the package class has a set of members but not all are in scope. Previously I've read a comment about stale symbols from previous run; symbol history, like human history, is a story of wreckage and detritus. Somehow I had never fully considered the state of the environment during incremental compilation.

File: test/files/neg/t12289/c_2.scala (hunk 1 above)
Comment (som-snytt) [reply test/files/neg/t12289/c_2.scala]: How about `-Ystop-before:jvm`. But a test for programmatic `outputDirs.outputs` would have to be `DirectTest`.

Comment (som-snytt) [reply test/files/neg/t12289/c_2.scala]: `-Yskip:jvm` is somehow cooler. Either works. Could stop earlier, as it's just a backstop.

**Comment by som-snytt** [comment]
Followed suggestions for better message, tweak test so `-d /tmp` never outputs, add test for the "multiple -d" case (not sure if that is used by current tooling) and handle that. Also handle shadowing of companion class or companion module, as pointed out in the review.

I agree with the previous comment that REPL's reminder that "companions must be defined together" is not optional. This synthetic shadowing is very tricky.

dotty has same behavior:
```
➜  scala git:(issue/12289-module-class-shadowing) scalac -d /tmp/sandbox test/files/neg/t12289/t_1.scala
➜  scala git:(issue/12289-module-class-shadowing) scalac -cp /tmp/sandbox -d /tmp/sandbox test/files/neg/t12289/c_2.scala
-- [E006] Not Found Error: test/files/neg/t12289/c_2.scala:6:18 ----------------------------------------------------------------------
6 |  class C extends T
  |                  ^
  |                  Not found: type T
  |
  | longer explanation available when compiling with `-explain`
1 error found
```

FSR I don't understand yet, the test under Vulpix says
```
-- [E006] Not Found Error: tests/neg/i20510/c_2.scala:6:18 -------------------------------------------------------------
6 |  class C extends T // error
                  ^
                  Not found: type T - did you mean T.type?
```

**Comment by som-snytt** [comment]
If I weren't so lazy, I'd help with why the spec build fails ~so often~ always. That was a problem in the old days, but why is it a problem in the new days?

image

**Review by lrytz — APPROVED** [review]
I hope this will prevent a lot of headscratching! Thanks.

**Comment by som-snytt** [comment]
Periodic reminder to get out the xpres spa scalp massager device. It's quite soothing.

**Comment by lrytz** [comment]
Some soothing might be needed because this PR broke on Windows... https://github.com/scala/scala/actions/runs/9369646735/job/25794666793

`+error: /tmp does not exist or is not a directory`

**Comment by lrytz** [comment]
> why the spec build fails

https://travis-ci.community/t/travis-focal-ubuntu-image-uses-now-expired-mongodb-4-4-package/14293/4

**Comment by som-snytt** [comment]
Well, you warned me about `/tmp`. I'll follow up now.

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala (hunk 3 above)
Comment (som-snytt) [reply src/compiler/scala/tools/nsc/typechecker/Typers.scala:5682]: 🤯

Comment (som-snytt) [reply src/compiler/scala/tools/nsc/typechecker/Typers.scala:5682]: returned here because t5639 about not finding implicit because of stale symbol

**Pull request merged**